**Project - Book Business**

**Objective of the project:**

We need to analyze the data. Data structure
- book information
- information about publishers
- information about the authors
- custom book reviews

**Initial data:**

Database containing tables: books, authors, publishers, ratings, and reviews.

**Tasks**
- Count how many books have been published since January 1, 2000;
- For each book, count the number of reviews and the average rating;
- Determine the publishing house that has released the largest number of books thicker than 50 pages - so you exclude brochures from the analysis;
- Identify the author with the highest average book rating - consider only books with 50 or more ratings;
- Calculate the average number of reviews from users who have given more than 50 ratings

**Note:**

The requests were made as part of a training project; connection to the database is possible only from the training environment provided by Yandex.Practice.

## Explore the tables - print the first rows

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
query_1 = '''
         SELECT *
         FROM books

        '''
books = pd.io.sql.read_sql(query_1, con = engine)
books.head(3)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [4]:
query_2 = '''
         SELECT *
         FROM authors

        '''
authors = pd.io.sql.read_sql(query_2, con = engine)
authors.head(3)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

In [5]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [6]:
query_3 = '''
         SELECT *
         FROM publishers

        '''
publishers = pd.io.sql.read_sql(query_3, con = engine)
publishers.head(3)

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

In [7]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [8]:
query_4 = '''
         SELECT *
         FROM ratings

        '''
ratings = pd.io.sql.read_sql(query_4, con = engine)
ratings.head(3)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

In [9]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [10]:
query_5 = '''
         SELECT *
         FROM reviews

        '''
reviews = pd.io.sql.read_sql(query_5, con = engine)
reviews.head(3)

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

In [11]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Conclusions:
- all data loaded
- no data type issues
- no passes

## Let's make one SQL query to solve each task

### Let's count how many books have been published since January 1, 2000

In [12]:
query_6 = '''
         SELECT COUNT(book_id) AS number_of_books
         FROM books
         WHERE publication_date > '2000-01-01'
        '''
number_of_books_after_01_01_2000 = pd.io.sql.read_sql(query_6, con = engine)
number_of_books_after_01_01_2000

number_of_books
0              819

since January 1, 2000, 819 books have been published

### Let's count for each book the number of reviews and the average rating

In [13]:
query_7 = '''
SELECT 
    books.book_id, 
    books.title, 
    COUNT(DISTINCT review_id) AS number_of_reviews,
    AVG (rating) AS avg_rating
FROM 
    books
    LEFT JOIN reviews ON reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY books.book_id
         ORDER BY avg_rating DESC, number_of_reviews DESC
        '''
number_of_reviews_and_rating = pd.io.sql.read_sql(query_7, con = engine)
number_of_reviews_and_rating


book_id                                              title  \
0         17                      A Dirty Job (Grim Reaper  #1)   
1        553            School's Out—Forever (Maximum Ride  #2)   
2        444       Moneyball: The Art of Winning an Unfair Game   
3         86      Arrows of the Queen (Heralds of Valdemar  #1)   
4        972  Wherever You Go  There You Are: Mindfulness Me...   
..       ...                                                ...   
995      915  The World Is Flat: A Brief History of the Twen...   
996      202                                      Drowning Ruth   
997      316                  His Excellency: George Washington   
998      371                                              Junky   
999      303                               Harvesting the Heart   

     number_of_reviews  avg_rating  
0                    4        5.00  
1                    3        5.00  
2                    3        5.00  
3                    2        5.00  
4                    2        5.00  
..                 ...         ...  
995                  3        2.25  
996                  3        2.00  
997                  2        2.00  
998                  2        2.00  
999                  2        1.50  

[1000 rows x 4 columns]

Conclusions: The biggest average rating is 5.0

### Let's find the publisher that has released the most books that are over 50 pages.

In [14]:
query_8 = '''
         SELECT publishers.publisher_id, publishers.publisher, COUNT(books.book_id) as number_of_books
         FROM publishers
         JOIN books ON publishers.publisher_id = books.publisher_id
         WHERE books.num_pages > 50
         GROUP BY publishers.publisher_id
         ORDER BY number_of_books DESC
         LIMIT 1
        '''
pd.io.sql.read_sql(query_8, con = engine)

publisher_id      publisher  number_of_books
0           212  Penguin Books               42

Conclusion: Penguin Books (publisher_id = 212) has published 42 books, the number of pages is more than 50.

### Let's find the author with the highest average book rating. For a book with 50 or more ratings

In [15]:
query_9 = '''
         SELECT authors.author_id, authors.author, AVG(avg_rating) AS avg_rating_total
         FROM authors
         JOIN books ON books.author_id = authors.author_id
         JOIN(
             SELECT books.book_id, books.title, COUNT(rating_id) AS number_of_ratings, AVG (rating) AS avg_rating
             FROM books
             JOIN ratings ON ratings.book_id = books.book_id
             GROUP BY books.book_id, books.title) AS subq ON subq.book_id = books.book_id
         WHERE number_of_ratings >= 50
         GROUP BY authors.author_id, authors.author
         ORDER BY avg_rating_total DESC
         LIMIT 1
        '''
pd.io.sql.read_sql(query_9, con = engine)

author_id                      author  avg_rating_total
0        236  J.K. Rowling/Mary GrandPré          4.283844

Conclusion: The author with the highest average book rating (among authors with 50 or more ratings) is J.K. Rowling/Mary GrandPré (author_id = 236), her average book rating is 4.28

### Let's find the average number of reviews from users who have given more than 50 ratings.

In [16]:
query_13 = '''
     SELECT AVG(count)
     FROM (
         SELECT reviews.username, COUNT(reviews.review_id)
         FROM reviews
         JOIN (
                 SELECT *
         FROM
         (SELECT ratings.username, COUNT(ratings.rating_id) as number_of_ratings
         FROM ratings
         GROUP BY ratings.username
         ORDER BY number_of_ratings DESC
         ) AS subq
         WHERE number_of_ratings > 50
         ) AS subq_2 ON subq_2.username = reviews.username
         GROUP BY reviews.username
         ) AS subq_3
     '''
pd.io.sql.read_sql(query_13, con = engine)

avg
0  24.333333

Conclusion: the average number of reviews by users who gave more than 50 ratings is 24.3